### ISSUE TITLE
Improve developer productivity for the rules that generates source files.

### ISSUE THREAD

**Issue by thesayyn** [bazelbuild/rules_nodejs#1574]
### Problem 
In recent my recent experiments I saw that some of the rules like `ts_proto_library` generate source files in `bazel-bin` directory that can not be indexed by the typescript language service.  

For instance `ts_proto_library` compiles `.proto` files to `.ts` files. However, those files cannot be indexed since they are in a directory that ignored by .gitignore or something similar in a basic workspace setup. 

When a rule generates source files instead of compiled files, eventually there will be a drawback that affects developer productivity in a bad manner since some of the basic language features like `code completion`, `diagnostics` etc won't provide information about those files.

### Workaround
As far as I see, most people add  `bazel-bin` directory to `rootDir` section of the `tsconfig.json` to work around this issue. In my case, it wasn't a good solution cause it was blowing up the `tsserver` process eventually.

The reason behind the crashes was out of memory errors obviously since the amount of the files indexed was multiplied by two. 

### Solution
I had been thinking about this problem and I found a scalable solution that might fix this problem once for all. As you might think, writing a `tsserver` plugin adds those files to `ts.Program` but not the one that adds all  `.ts` files in `bazel-bin` directory to `ts.Program`. 

If we do add all files that have `.ts`  suffix, the `tsserver` will blow up in the end cause it won't be any different than adding `bazel-bin` directory to `rootDirs` array. Even if it does not blow up, it will consume all the resources your computer has. 
Therefore doubling the amount of the files that will be indexed by `tsserver` is something that we should avoid.

In this case, what we need to do is find the missing files individually and add them `ts.Program` to get diagnostics about those files. 

In order to do that, we have to find all the bazel targets which generate `.ts` files in the ` bazel-bin` directory and looks like the `bazel aquery` command does exactly this. 
Through `aquery` command we can analyze all the targets in the workspace and find the `.ts` generating targets and generated files.

In conclusion, writing a `tsserver` plugin seems possible and worth giving a try. By taking the approach above, we can keep memory and CPU usage low since we are not going to add all files.

Note: I started to work on a plugin that is going to do this. Before I go any further, I wanted to confirm and know your thoughts about this

**Comment by thesayyn** [comment]
cc @alexeagle @gregmagolan

**Comment by Toxicable** [comment]
Im not certain that writing a plugin is the correct approach here.

Before going down such a route, I'd first suggest debugging tsserver a little bit.
Do we have some numbers on what causes the tsserver to run out of memory?
Can we get a list of files it's watching? Maybe it's picking up `.js` files

It's possible we could improve `tsserver` by upstreaming improvements or configuring it in a way that better suites `rules_nodejs`

**Comment by thesayyn** [comment]
I'm sure about that it is not picking up the js files, only d.ts files which already indexed in source directory. In the company that I am working we have a big codebase which uses bazel. To be honest, It was the last time the tsserver got crashed after a few rebuilds. But yes i will debug it next time and write the results here.

**Comment by joeljeske** [comment]
Does this problem account for the issue where TS files are generated from Bazel and other source files import those generated files directly?

I’ve had trouble getting the IDE / TSServer to recognize those generated files. It works during Bazel compile since all the proper files exists in the build directory, but they do not exist in the source tree making the DX less than ideal

**Comment by thesayyn** [comment]
This solution takes those files into account.
However there will be a problem after the sandbox had cleaned since the generated files won't be there and as far as I know bazel query command generates the generated files before it returns the result of the query but I'm not quite sure. So if it does, this solution will work even after the sandbox had been cleaned.

**Comment by joeljeske** [comment]
Ok that makes sense. To be clear, are you talking about the sandbox that is cleaned on every sandboxed execution? Or are you talking about the Bazel-out which would be cleaned when running “bazel clean”

@Toxicable, are you suggesting that instead of a tsserver plugin, that a direct PR would be made to TSServer to be able to gather files outside the normal src tree?

**Comment by thesayyn** [comment]
I am talking about bazel-out directory.

**Comment by thesayyn** [comment]
I think this is still an issue that should be solved.

**Comment by gregmagolan** [comment]
No longer in scope for rules_nodejs which only supplies the Node.js toolchain as of [v6.0.0](https://github.com/bazelbuild/rules_nodejs/releases/tag/v6.0.0-beta0).

Downstream canonical JavaScript + Node.js ruleset is now https://github.com/aspect-build/rules_js.
